designs/2021-suppression-support/README.md

```diff
@@ -0,0 +1,302 @@
+a = 1;
+/* eslint-enable no-undef */
+```
+```javascript
+console.log("foo"); // eslint-disable-line no-console -- Justification example
+```
+```javascript
+// eslint-disable-next-line no-console -- Justification example
+console.log("bar");
+```
+
+Then we may get:
+
+```json
+{"kind": "inSource", "justification": "Justification example"}
+```
+
+If a violation is inline-suppressed multiple times, all suppression entries will be recorded, according to [suppressions property](https://docs.oasis-open.org/sarif/sarif/v2.0/csprd02/sarif-v2.0-csprd02.html#_Toc10127852). So `unusedDisableDirectives` might always be empty.
+
+For external suppression (or configuration files in ESLint), there are no good places to put a justification. We may put something general as a justification, like "Globally disabled".
+
+For example:
+
+```json
+{"kind": "external", "justification": "Globally disabled"}
+```
+
+If a line is suppressed by multiple rules, a suppression list would be preferred to gather all suppression information:
+
+```json
+suppressions: [
+  {"kind": "inSource", "justification": "Justification message 1"},
+  {"kind": "inSource", "justification": "Justification message 2"},
+  {"kind": "external", "justification": "Globally disabled"}
+]
+```
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+### Goal
+
+1.	Keep the current behavior of ESLint, i.e., errors/warnings will not be shown on the command line if the related rules are disabled, unless `--no-inline-config` is used.
+2.	When the flag `--output-suppression` is used, the violations (problems) that disabled by inline comments will not be removed.
+3.	Add suppression info in the output of ESLint
+    1. For inSource suppressions, we generate suppression info where `kind` is `inSource` and `justification` is the description in the directive comments (words after two or more consecutive dashes mentioned above).
+    
+        For example:
+        
+        ```javascript
+        /* eslint-disable no-undef -- Justification example */
+        a = 1;
+        /* eslint-enable no-undef */
+        ```
+        ```javascript
+        console.log("foo"); // eslint-disable-line no-console -- Justification example
+        ```
+        ```javascript
+        // eslint-disable-next-line no-console -- Justification example
+        console.log("bar");
+        ```
+
+        Output:
+
+        ```json
+        {"kind": "inSource", "justification": "Justification example"}
+        ```
+    
+    2. For external suppressions, we generate suppression info where `kind` is `external` and `justification` is a certain sentence such as "Globally disabled".
```

Comment:
To clarify. 
```suggestion
    2. When rules are disabled in a config file or on the command line, we generate suppression info where `kind` is `external` and `justification` is a certain sentence such as "Globally disabled".
```

The phrase “external suppression” isn’t part of ESLint terminology.

Reply:
Sorry for the confusion. "External suppression" is disabling rules inside of configuration files, which is the opposite of "inSource suppression", i.e., disabling rules with inline directive comments.
I've added explanations for all "external suppression" mentioned in the doc. Welcome to point out more terminology issues.